src/i_pcsound.c

```diff
@@ -0,0 +1,402 @@
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// Mixer function that does the PC speaker emulation
+
+static ALsizei AL_APIENTRY BufferCallback(void *userptr, void *data, ALsizei size)
+{
+    int16_t *leftptr;
+    int16_t *rightptr;
+    int16_t this_value;
+    int frequency;
+    int i;
+    int nsamples;
+
+    // Number of samples is quadrupled, because of 16-bit and stereo
+
+    nsamples = size / 4;
+
+    leftptr = (int16_t *) data;
+    rightptr = ((int16_t *) data) + 1;
+
+    // Fill the output buffer
+
+    for (i = 0; i < nsamples; ++i)
+    {
+        // Has this sound expired? If so, invoke the callback to get the next
+        // frequency.
+
+        while (current_remaining == 0) 
+        {
+            // Get the next frequency to play
+
+            GetFreq(&current_remaining, &frequency);
+
+            if (current_freq != frequency)
+            {
+                current_freq = frequency;
+                phase_offset = 0;
+            }
+
+            current_remaining = (current_remaining * mixing_freq) / 1000;
+        }
+
+        // Set the value for this sample.
+
+        if (current_freq == 0)
+        {
+            // Silence
+
+            this_value = 0;
+        }
+        else
+        {
+            int frac;
+
+            // Determine whether we are at a peak or trough in the current
+            // sound.  Multiply by 2 so that frac % 2 will give 0 or 1 
+            // depending on whether we are at a peak or trough.
+
+            frac = (phase_offset * current_freq * 2) / mixing_freq;
+
+            if ((frac % 2) == 0) 
+            {
+                this_value = SQUARE_WAVE_AMP;
+            }
+            else
+            {
+                this_value = -SQUARE_WAVE_AMP;
+            }
+
+            ++phase_offset;
+        }
+
+        --current_remaining;
+
+        // Use the same value for the left and right channels.
+
+        *leftptr = this_value;
+        *rightptr = this_value;
+
+        leftptr += 2;
+        rightptr += 2;
+    }
+
+    return size;
+}
+
+static boolean I_PCS_InitSound(void)
+{
+    if (!I_OAL_InitSound())
+    {
+        return false;
+    }
+
+    mixing_freq = SND_SAMPLERATE;
+
+    current_freq = 0;
+    current_remaining = 0;
+
+    I_OAL_RegisterCallback(BufferCallback);
+
+    sound_lock = SDL_CreateMutex();
+
+    return true;
+}
+
+static boolean I_PCS_CacheSound(sfxinfo_t *sfx)
+{
+    if (IsDisabledSound(sfx))
+    {
+        return false;
+    }
+
+    return CachePCSLump(sfx);
+}
+
+static void I_PCS_ShutdownSound(void)
+{
+    int i;
+
+    for (i = 0; i < num_sfx; ++i)
+    {
+        S_sfx[i].lumpnum = -1;
+    }
+
+    I_OAL_UnregisterCallback();
+    I_OAL_ShutdownSound();
+}
+
+static boolean I_PCS_AdjustSoundParams(const mobj_t *listener, const mobj_t *source,
+                                      int chanvol, int *vol, int *sep, int *pri)
+{
+    return true;
```

Comment:
Although not strictly faithful to the original implementation, I'd prefer if we could adjust the volume for the effects.

Reply:
Fixed in https://github.com/fabiangreffrath/woof/pull/1200/commits/e54de3597af3b36ba0eca36dc16c301b2266cd48